### PR TITLE
add salt cloud support for block device tagging or root devices and such

### PR DESCRIPTION
**Pull request by thetoolsmith**
This pull request adds the ability to set block device tags in salt cloud configuration for root volumes and such. 

Currently, it seems that this is not supported in salt-cloud. 
ref: https://github.com/saltstack/salt/issues/47097

This change allows for configuration of block_device_mappings in cloud configuration to support a tag dictionary like so:

      block_device_mappings:
        - DeviceName: /dev/sda1
           Ebs.VolumeSize: 10
           Ebs.VolumeType: gp2
           tag:    
             tag1: value
             tag2: value
        - DeviceName: /dev/sdb
           Ebs.VolumeSize: 35
           Ebs.VolumeType: gp2
           tag:    
             tag3: value
             tag4: value

### Tests written?
No

### Commits signed with GPG?
No

Additional notes:
I also have another implementation that does not use the existing block_device_mappings configuration. Instead it would add a new cloud configuration block_device_tags. However, this change separates the mappings from the tags and might not be ideal. 
I'm thinking this pr is the best approach, but will look for feedback from the core Salt team.

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/latest/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Comment by thetoolsmith** [comment]
build kicked off before I committed the corrected code.

**Comment by rallytime** [comment]
@thetoolsmith Can you fix these lint errors?

https://jenkinsci.saltstack.com/job/pr-lint/job/PR-48716/2/warnings52Result/

**Comment by thetoolsmith** [comment]
yeah, sorry about that. I pasted the code into the salt branch from my test system and missed the closing bracket. should have re-checked the code. 
fix is committed.

File: salt/cloud/clouds/ec2.py [hunk 1]
```diff
@@ -2856,6 +2862,53 @@ def create(vm_=None, call=None):
...
+    blockdevicemappings_holder = block_device_mappings(vm_)
+    if blockdevicemappings_holder:
+        for _bd in blockdevicemappings_holder:
+            if 'tag' in _bd:
+                ex_blockdevicetags[_bd['DeviceName']] = _bd['tag']
+
+    block_device_volume_id_map = {}
+
+    if ex_blockdevicetags:
+        for k, v in six.iteritems(ret['blockDeviceMapping']):
```
Comment (rallytime) [inline salt/cloud/clouds/ec2.py]: Can you switch these variables (and others below) to use 3 or more letters? That would be nice and match our [Style Guide](https://docs.saltstack.com/en/latest/topics/development/conventions/style.html#variables) rules.

File: salt/cloud/clouds/ec2.py [hunk 2]
```diff
@@ -2856,6 +2862,53 @@ def create(vm_=None, call=None):
...
+            __utils__['cloud.fire_event'](
+                'event',
+                'setting tags',
+                'salt/cloud/block_volume_{0}/tagging'.format(str(volid)),
+                args={'tags': tags},
+                sock_dir=__opts__['sock_dir'],
+                transport=__opts__['transport']
+            )
+
+            salt.utils.cloud.wait_for_fun(
```
Comment (rallytime) [inline salt/cloud/clouds/ec2.py]: This can be the `__utils__['cloud.x']` notation like you have above.

**Review by rallytime — CHANGES_REQUESTED** [review]
In addition to the inline requests, can you also document this new behavior?

Thanks @thetoolsmith!

File: salt/cloud/clouds/ec2.py (hunk 1 above)
Comment (thetoolsmith) [reply salt/cloud/clouds/ec2.py]: sure thing

**Comment by thetoolsmith** [comment]
changes have been comitted

**Review by rallytime — APPROVED** [review]

**Review by gtmanfred — APPROVED** [review]

**Pull request merged**